File: examples/cv-scanner/python/src/store.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path

from .contract import CvScanResult

DB_PATH = Path(__file__).parent.parent / "scans.json"


def _load() -> list[dict]:
    if not DB_PATH.exists():
        return []
    return json.loads(DB_PATH.read_text())


def _save(scans: list[dict]) -> None:
    DB_PATH.write_text(json.dumps(scans, indent=2))


def add_scan(scan: CvScanResult, file: str) -> dict:
    scans = _load()
    record = {
        **scan.model_dump(),
        "id": len(scans) + 1,
        "scanned_at": datetime.now(timezone.utc).isoformat(),
        "file": file,
    }

    print(
        f"[Boundary] Saving scan #{record['id']}: "
        f"{record.get('full_name') or 'Unknown'} — "
        f"{record['extraction_quality']} "
        f"({len(record.get('experience') or [])} job(s), "
        f"{len(record.get('skills') or [])} skill(s))"
    )
    scans.append(record)
    _save(scans)
    print(
        f"[Boundary] Scan #{record['id']} written to store"
        f" (total: {len(scans)} record(s))"
    )
    return record


def list_scans() -> list[dict]:
    return _load()

File: examples/cv-scanner/node/src/store.ts
import fs from "fs";
import path from "path";
import { type CvScanResult } from "./contract.js";

const DB_PATH = path.resolve("scans.json");

export interface StoredScan extends CvScanResult {
  id: number;
  scannedAt: string;
  file: string;
}

function load(): StoredScan[] {
  if (!fs.existsSync(DB_PATH)) return [];
  return JSON.parse(fs.readFileSync(DB_PATH, "utf-8")) as StoredScan[];
}

function save(scans: StoredScan[]): void {
  fs.writeFileSync(DB_PATH, JSON.stringify(scans, null, 2));
}

export function addScan(result: CvScanResult, file: string): StoredScan {
  const scans = load();
  const scan: StoredScan = {
    ...result,
    id: scans.length + 1,
    scannedAt: new Date().toISOString(),
    file,
  };

  console.log(
    `[Boundary] Saving scan #${scan.id}: ${scan.full_name ?? "Unknown"} — ` +
      `${scan.extraction_quality} (${scan.experience.length} job(s), ${scan.skills.length} skill(s))`,
  );
  scans.push(scan);
  save(scans);
  console.log(`[Boundary] Scan #${scan.id} written to store (total: ${scans.length} record(s))`);
  return scan;
}

export function listScans(): StoredScan[] {
  return load();
}
